server/vulnerabilities/oval/parser.go

```diff
@@ -0,0 +1,164 @@
+package oval
+
+import (
+	"encoding/json"
+	"encoding/xml"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	oval_input "github.com/fleetdm/fleet/v4/server/vulnerabilities/oval/input"
+	oval_parsed "github.com/fleetdm/fleet/v4/server/vulnerabilities/oval/parsed"
+)
+
+func parseDefinitions(inputFile string, outputFile string) error {
+	r, err := os.Open(inputFile)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+
+	xmlResult, err := parseXML(r)
+	if err != nil {
+		return err
+	}
+
+	result, err := mapResult(xmlResult)
+	if err != nil {
+		return err
+	}
+
+	payload, err := json.Marshal(result)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(outputFile, payload, 0o644)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func mapResult(xmlResult *oval_input.UbuntuResultXML) (*oval_parsed.UbuntuResult, error) {
+	r := oval_parsed.NewResult()
+
+	staToTst := make(map[string][]int)
+	objToTst := make(map[string][]int)
+
+	for _, d := range xmlResult.Definitions {
+		if len(d.CVEs) > 0 {
+			def, err := mapDefinition(d)
+			if err != nil {
+				return nil, err
+			}
+			r.AddDefinition(*def)
+		}
+	}
+
+	for _, t := range xmlResult.PackageTests {
+		id, tst, err := mapPackageTest(t)
+		if err != nil {
+			return nil, err
+		}
+
+		objToTst[t.Object.Id] = append(objToTst[t.Object.Id], id)
+		for _, sta := range t.States {
+			staToTst[sta.Id] = append(staToTst[sta.Id], id)
+		}
+
+		r.AddPackageTest(id, tst)
+	}
+
+	for _, o := range xmlResult.PackageObjects {
+		obj, err := mapPackageObject(o, xmlResult.Variables)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, tId := range objToTst[o.Id] {
+			t, ok := r.PackageTests[tId]
+			if ok {
+				t.Objects = obj
+			} else {
+				return nil, fmt.Errorf("test not found: %d", tId)
+			}
+		}
+	}
+
+	for _, s := range xmlResult.PackageStates {
+		sta, err := mapPackageState(s)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, tId := range staToTst[s.Id] {
+			t, ok := r.PackageTests[tId]
+			if ok {
+				t.States = sta
+			} else {
+				return nil, fmt.Errorf("test not found: %d", tId)
+			}
+		}
+	}
+	return r, nil
+}
+
+func parseXML(reader io.Reader) (*oval_input.UbuntuResultXML, error) {
```

Comment:
I'm curious why you had to parse the token stream manually here? Golang does support parsing xml using xml tags similar to json. Here is a package that does just that for oval https://github.com/quay/goval-parser/blob/69102cd5c9205834e5722306e05589c13b5d1c8c/oval/types.go#L19

Reply:
I'm using XML tags, the difference is in the way the doc is un-marshaled/parsed. Instead of loading the whole document into memory at once (like in the library you linked), I'm streaming through the doc and parsing it piece by piece (or rather token by token). The former is usually faster,  the latter is more memory efficient. Since the OVAL files are relatively large (some can take up to 20 Mb) and the parsing is performed asynchronously (by a 'cron' job) I choose to optimize for memory instead of time.

Reply:
Any other reason to implement our own parser? Just curious, this is great work.

Reply:
Also asking because 20MB of memory doesn't sound like an issue (currently vulnerability processing requires > 1 GBs of memory IIRC).

Reply:
Nah, that's it

Reply:
Marking this as resolved, if you guys feel like we need to further discuss this, please unresolve it